### PR TITLE
.github: Add differential-shellcheck workflow for shell script analysis

### PR DESCRIPTION
**Pull request by tchaikov**
Introduce a new GitHub workflow to run shellcheck on changed shell scripts. This workflow automatically detect and highlight potential shell script issues in pull requests. This change is a follow-up to commit 0700b322 which fixed an undefined variable issue in `install.sh`. It intends to leverage static analysis to improve script quality and catch potential errors early.

Shellcheck will now:
- Analyze all shell scripts modified in pull requests
- Provide inline comments with specific issue details
- Help prevent similar variable-related mistakes in the future

See also
https://github.com/redhat-plumbers-in-action/differential-shellcheck

---

this is an improvement of the workflow to identify potential issues in shell scripts in newly introduced changes, hence no need to backport.

### CODE REVIEW

**Comment by tchaikov** [comment]
i tested this change in my own github repo by sending a pull request reverting https://github.com/scylladb/scylladb/commit/0700b322b8f9a9654183276828a59a1f583e04e9 , see https://github.com/tchaikov/scylladb/pull/12. the workflow identified the issue as expected.

**Comment by mykaul** [comment]
@gmizrahi , @d-helios - probably useful to implement in your respective repos as well.

**Comment by gmizrahi** [comment]
> @gmizrahi , @d-helios @yaronkaikov - probably useful to implement in your respective repos as well.

we are adding something similar to SQL Schema changes - https://github.com/scylladb/siren/issues/12540

cc @ylebi

File: .github/workflows/differential-shellcheck.yaml [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+on:
+  push:
+    branches:
+      - master
+  pull_request:
+    branches:
+      - master
+
+permissions:
+  contents: read
```
Comment (yaronkaikov) [inline .github/workflows/differential-shellcheck.yaml:14]: I think you can change this to `write` and drop lines 20-21?

Comment (tchaikov) [reply .github/workflows/differential-shellcheck.yaml:14]: You raise an interesting point about lines 20-21. However, these permissions serve different purposes according to the GitHub documentation:

The recommended configuration distinguishes between two key permission scopes:
- `contents: read` allows viewing repository contents
- `actions: write` enables modifying workflow runs and their status

These aren't redundant but have distinct functions. As detailed in the [GitHub docs](https://docs.github.com/en/enterprise-server@3.11/actions/writing-workflows/choosing-what-your-workflow-does/controlling-permissions-for-github_-token), each permission grants specific capabilities.

quote from the linked document:

| Permission | Allows an action using GITHUB_TOKEN to |
| --- | --- |
| `content` |  Work with the contents of the repository. For example, contents: read permits an action to list the commits, and contents: write allows the action to create a release. For more information, see "[Permissions required for GitHub Apps](https://docs.github.com/en/enterprise-server@3.11/rest/overview/permissions-required-for-github-apps?apiVersion=2022-11-28#repository-permissions-for-contents)."
| `security-events` | Work with GitHub code scanning and Dependabot alerts. For example, security-events: read permits an action to list the Dependabot alerts for the repository, and security-events: write allows an action to update the status of a code scanning alert. For more information, see "[Repository permissions for 'Code scanning alerts'](https://docs.github.com/en/enterprise-server@3.11/rest/overview/permissions-required-for-github-apps?apiVersion=2022-11-28#repository-permissions-for-code-scanning-alerts)" and "[Repository permissions for 'Dependabot alerts'](https://docs.github.com/en/enterprise-server@3.11/rest/overview/permissions-required-for-github-apps?apiVersion=2022-11-28#repository-permissions-for-dependabot-alerts)" in "Permissions required for GitHub Apps." |

also, I'm implementing the recommended configuration from the Differential ShellCheck GitHub Action, with some modifications tailored to our specific project requirements based on https://github.com/redhat-plumbers-in-action/differential-shellcheck?tab=readme-ov-file#usage .

**Review by yaronkaikov — APPROVED** [review]

File: .github/workflows/differential-shellcheck.yaml [hunk 2]
```diff
@@ -0,0 +1,32 @@
+---
+# https://github.com/redhat-plumbers-in-action/differential-shellcheck#readme
+
+name: Differential ShellCheck
+on:
+  push:
+    branches:
+      - master
```
Comment (yaronkaikov) [inline .github/workflows/differential-shellcheck.yaml:8]: @tchaikov I know this one is already queued, @Annamikhlin raised a good point, why do you need to run it during the push to `master` if you are running during the PR opening?

Comment (tchaikov) [reply .github/workflows/differential-shellcheck.yaml:8]: @yaronkaikov this is indeed a good question. i pondered on it for a while. i think it depends on the development model: we merge pull requests into master in batch. so there are chances that two pull requests are sane if we look at them individually, but their combination could cause a bug.  if we trigger this workflow with push to master, we would be able to catch this issue.

what do you think?

Comment (yaronkaikov) [reply .github/workflows/differential-shellcheck.yaml:8]: I assume it;s an end case, but if it will cover it, i would leave it as is

**Pull request closed without merging**